[MIR] Get rid of that nasty unit_ty temporary lval

nagisa opened the pull request:
Get rid of that nasty unit_ty temporary variable created just because it might be handy to have one around, when in reality it isn’t really that useful at all.

r? @nikomatsakis 

Fixes https://github.com/rust-lang/rust/issues/30637

nikomatsakis commented on src/librustc_mir/build/expr/into.rs:
```diff
@@ -156,11 +156,16 @@ impl<'a,'tcx> Builder<'a,'tcx> {
-                    let unit_temp = this.unit_temp.clone();
-                    let body_block_end = unpack!(this.into(&unit_temp, body_block, body));
+                    // FIXME(#30636): this should not create or request any sort of destination at
+                    // all.
```
I'm not sure I agree with this FIXME. In trans-land, we have some special cases for "ignoring" the result and it's kind of a pain. I guess though that we could have a special mode here *just for blocks*, maybe.

nikomatsakis replied:
Also, I'm mildly worried that writing to `destination` here will yield incorrect results for the initialization analysis, once we implement that. Seems like something like:

```
let x;

```
x = while foo { ... /* try to use `x` in here */ }
```

might succeed when it shouldn't. But hmm that particular example I guess would not, since there is at least one iteration (the first one) where the body has not executed, and thus has not written into `destination`. Hmm, I guess it's ok.

nikomatsakis commented on src/librustc_mir/build/block.rs:
```diff
@@ -19,11 +19,18 @@ impl<'a,'tcx> Builder<'a,'tcx> {
             unpack!(block = this.stmts(block, stmts));
-            this.into(destination, block, expr)
+            match expr {
+                Some(expr) => this.into(destination, block, expr),
```
Do we still need the impl of `EvalInto` for `Option<ExprRef<'tcx>>`? I suspect not. In any case I think perhaps we ought to remove it, since it's ambiguous what should happen on `None`.

nikomatsakis commented on src/librustc_mir/build/expr/into.rs:
```diff
@@ -157,11 +157,8 @@ impl<'a,'tcx> Builder<'a,'tcx> {
                     }
 
                     // execute the body, branching back to the test
```
we should add at least a comment here explaining that we are (ab)using `destination` to serve as a dummy location

nikomatsakis replied:
and that we think this is ok because (a) the type of `destination` should always be `()` and (b) the destination will still be considered uninitialized for the first iteration, so it should not cause anything to type-check that wouldn't otherwise

nikomatsakis commented:
r+ modulo the 2 nits

nagisa commented:
Addressing the comments resulted in more changes (see ExprKind::If and ExprKind::Return (similar bugfix to the block one)), so I’m requesting a re-review.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 3692ab6 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 3692ab673e45488ef0d276618a12982964e95c9e -->

bors commented:
:hourglass: Testing commit 3692ab6 with merge d6cb279...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7654), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7673), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/925), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7609), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7627), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7611), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/836), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2785), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7679), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7675), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7639), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2651), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2665), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2667), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2645), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1894), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2649)
